File: src/Java/Parser.java
package Java;

import Java.ArbolSincatico.*;

import java.util.ArrayList;
import java.util.Hashtable;

public final class Parser {
    private Tokens tkn;
    private Token thisToken;
    private ArrayList<String> estatutos;
    public ArrayList<Declarax> declaraciones = new ArrayList<Declarax>();
    public ArrayList<Statx> statements = new ArrayList<Statx>();
    Hashtable<String, String> tabladesimolos = new Hashtable<String, String>();
    private Scanner scan;
    Programax p;
    private void advance() {
        thisToken = scan.getToken();
        tkn = thisToken.getTipo();
    }

    private void eat(Tokens t) {
        if (thisToken.getClasificacion() == -2) // Si el token es Fin del archivo
        {
            estatutos.add("Token Esperado " + t.getCad() + " Token Recibido <EOF>");
            estatutos.add("Error..... Se esperaba " + t.getCad() + " no <EOF>$");
            return;
        }
        estatutos.add("Token Esperado " + t.getCad() + " Token Recibido " + tkn.getCad()+" en la linea "+thisToken.getLinea());
        if (tkn == t)
            advance();
        else
            estatutos.add("Error..... Se esperaba " + t.getCad() + " no " + tkn.getCad() +" en la linea "+thisToken.getLinea()+ "$");
    }

    public final void program() {
        if (tkn == Tokens.CLASS) {
            eat(Tokens.CLASS);
            eat(Tokens.ID);
            varDeclaration();
            eat(Tokens.LBRACE);
            while (tkn == Tokens.DO || tkn == Tokens.SIR || tkn == Tokens.ID || tkn == Tokens.LBRACE) {
                statement();
            }
            eat(Tokens.RBRACE);
            p = new Programax(declaraciones, statements);
        } else {
            estatutos.add("Error$");
        }

    }

    private Statx statement() {
        Expx ex;
        if (tkn == Tokens.LBRACE) {
            eat(Tokens.LBRACE);
            while (tkn == Tokens.DO || tkn == Tokens.SIR || tkn == Tokens.ID) {
                statement();
            }
            eat(Tokens.RBRACE);
            return null;
        } else if (tkn == Tokens.DO) {
            Statx s ;
            Dox dox;
            eat(Tokens.DO);
            do {
                s = statement();
            }while (tkn == Tokens.DO || tkn == Tokens.SIR || tkn == Tokens.ID || tkn == Tokens.LBRACE);
            eat(Tokens.UNTIL);
            eat(Tokens.LPAREN);
            ex = expresion();
            eat(Tokens.RPARENT);
            dox = new Dox(s, ex);
            statements.add(dox);
            return dox;
        } else if (tkn == Tokens.SIR) {
            Readx read;
            eat(Tokens.SIR);
            eat(Tokens.LPAREN);
            ex = expresion();
            eat(Tokens.RPARENT);
            eat(Tokens.SEMI);
            read = new Readx(ex);
            statements.add(read);
           return read;
        } else if (tkn == Tokens.ID) {
            Asignax as;
            Idx idx = new Idx(thisToken.getValor(), new Typex(tabladesimolos.get(thisToken.getValor())));
            eat(Tokens.ID);
            eat(Tokens.EQ);
            ex = expresion();
            eat(Tokens.SEMI);
            as = new Asignax(idx, ex);
            statements.add(as);
            return as;
        }
        return null;
    }

    public final Expx expresion() {
        String tipo;
        Idx idx1;
        Idx idx2;
        if (tkn == Tokens.ID) {
            tipo = tabladesimolos.get(thisToken.getValor());
            idx1 = new Idx(thisToken.getValor(), new Typex(tipo));
            eat(Tokens.ID);
            if (tkn == Tokens.MIN) {
                eat(Tokens.MIN);
                tipo = tabladesimolos.get(thisToken.getValor());
                idx2 = new Idx(thisToken.getValor(), new Typex(tipo));
                eat(Tokens.ID);
                return new Comparax(idx1, idx2);
            } else if (tkn == Tokens.PLUS) {
                eat(Tokens.PLUS);
                tipo = tabladesimolos.get(thisToken.getValor());
                idx2 = new Idx(thisToken.getValor(), new Typex(tipo));
                eat(Tokens.ID);
                return new Sumax(idx1, idx2);
            } else if (tkn == Tokens.MINUS) {
                eat(Tokens.MINUS);
                tipo = tabladesimolos.get(thisToken.getValor());
                idx2 = new Idx(thisToken.getValor(), new Typex(tipo));
                eat(Tokens.ID);
                return new Multiplicax(idx1, idx2);
            } else if (tkn == Tokens.ASTER) {
                eat(Tokens.ASTER);
                tipo = tabladesimolos.get(thisToken.getValor());
                idx2 = new Idx(thisToken.getValor(), new Typex(tipo));
                eat(Tokens.ID);
                return new Restax(idx1, idx2);
            }
        } else if (tkn == Tokens.FALSE) {
            eat(Tokens.FALSE);
            return new Idx("", new Typex("Boolean"));
        } else if (tkn == Tokens.TRUE) {
            eat(Tokens.TRUE);
            return new Idx("", new Typex("Boolean"));
        } else if (tkn == Tokens.NFLOA) {
            eat(Tokens.NFLOA);
            return new Idx("", new Typex("Float"));
        } else if (tkn == Tokens.NINTE) {
            eat(Tokens.NINTE);
            return new Idx("", new Typex("Integer"));
        } else {
            estatutos.add("Error... falta una expresion en la linea "+thisToken.getLinea()+"$");
        }
        return null;
    }

    public final ArrayList<String> dameSalida() {
        return estatutos;
    }

    private final void varDeclaration() {
        if (tkn == Tokens.BOOLEAN) {
            eat(Tokens.BOOLEAN);
            String var = thisToken.getValor();
            eat(Tokens.ID);
            eat(Tokens.SEMI);
            tabladesimolos.put(var, "Boolean");
            declaraciones.add(new Declarax(var, new Typex("Boolean")));
            varDeclaration();
        } else if (tkn == Tokens.INT) {
            eat(Tokens.INT);
            String var = thisToken.getValor();
            eat(Tokens.ID);
            eat(Tokens.SEMI);
            tabladesimolos.put(var, "Int");
            declaraciones.add(new Declarax(var, new Typex("Int")));
            varDeclaration();
        } else if (tkn == Tokens.FLOAT) {
            eat(Tokens.FLOAT);
            String var = thisToken.getValor();
            eat(Tokens.ID);
            eat(Tokens.SEMI);
            tabladesimolos.put(var, "Float");
            declaraciones.add(new Declarax(var, new Typex("Float")));
            varDeclaration();
        }

    }

    public Parser(Scanner scan) {
        this.scan = scan;
        estatutos = new ArrayList<String>();
        thisToken = scan.getToken();
        tkn = thisToken.getTipo();
        estatutos.add("Iniciando el parsing");
    }
}

File: src/Parser.kt

import javax.swing.JOptionPane

class Parser(private var scan:Scanner){
    private var tkn: Int =0
    private var estatutos = ArrayList<String>()
    init {
        tkn = scan.getToken().tipo
        estatutos.add("Iniciando el parsing")
    }
    fun advance(){tkn = scan.getToken().tipo }
    fun eat(t:Int){
        if(tkn == -2)
            return
        estatutos.add("Token Esperado " + Tokens.values()[t].name + " Token Recibido " +  Tokens.values()[tkn].name)
         if(tkn == t) {
             advance();
             return
         }
        JOptionPane.showMessageDialog(null,"Error..... Se esperaba " + Tokens.values()[t].name+ " no " + Tokens.values()[tkn].name);
            estatutos.add("Error..... Se esperaba " + Tokens.values()[t].name+ " no " + Tokens.values()[tkn].name+"$");
    }
    fun program(){
        when(tkn){
            Tokens.CLASS.ordinal->{
                eat(Tokens.CLASS.ordinal)
                eat(Tokens.ID.ordinal)
                varDeclaration()
                eat(Tokens.LBRACE.ordinal)
                while(tkn == Tokens.DO.ordinal|| tkn == Tokens.SIR.ordinal ||tkn == Tokens.ID.ordinal ||tkn == Tokens.LBRACE.ordinal)
                    statement()
                eat(Tokens.RBRACE.ordinal)
            }
            else-> estatutos.add("Error")
        }
    }
    fun statement(){
        when(tkn) {
            Tokens.LBRACE.ordinal->{
                eat(Tokens.LBRACE.ordinal)
                while(tkn == Tokens.DO.ordinal|| tkn == Tokens.SIR.ordinal ||tkn == Tokens.ID.ordinal)
                    statement()
                eat(Tokens.RBRACE.ordinal)
            }
            Tokens.DO.ordinal-> {
                eat(Tokens.DO.ordinal)
                statement()
                eat(Tokens.UNTIL.ordinal)
                eat(Tokens.LPAREN.ordinal)
                expresion()
                eat(Tokens.RPARENT.ordinal)
            }
            Tokens.SIR.ordinal->{
                eat(Tokens.SIR.ordinal)
                eat(Tokens.LPAREN.ordinal)
                expresion()
                eat(Tokens.RPARENT.ordinal)
                eat(Tokens.SEMI.ordinal)
            }
            Tokens.ID.ordinal->{
                eat(Tokens.ID.ordinal)
                eat(Tokens.EQ.ordinal)
                expresion()
                eat(Tokens.SEMI.ordinal)
            }
        }
    }
    fun expresion(){
        when(tkn){
            Tokens.ID.ordinal->{
                eat(Tokens.ID.ordinal)
                when(tkn){
                    Tokens.MIN.ordinal->{
                        eat(Tokens.MIN.ordinal)
                        eat(Tokens.ID.ordinal)
                    }
                    Tokens.PLUS.ordinal->{
                        eat(Tokens.PLUS.ordinal)
                        eat(Tokens.ID.ordinal)
                    }
                    Tokens.MINUS.ordinal->{
                        eat(Tokens.MINUS.ordinal)
                        eat(Tokens.ID.ordinal)
                    }
                    Tokens.ASTER.ordinal->{
                        eat(Tokens.ASTER.ordinal)
                        eat(Tokens.ID.ordinal)
                    }
                }
            }
            Tokens.FALSE.ordinal-> eat(Tokens.FALSE.ordinal)
            Tokens.TRUE.ordinal-> eat(Tokens.TRUE.ordinal)
            Tokens.NFLOA.ordinal-> eat(Tokens.NFLOA.ordinal)
            Tokens.NINTE.ordinal-> eat(Tokens.NINTE.ordinal)
            else -> estatutos.add("Error... falta una expresion$")
        }
    }
    fun dameSalida() = estatutos
    fun varDeclaration(){
        when(tkn){
            Tokens.BOOLEAN.ordinal->{
                eat(Tokens.BOOLEAN.ordinal)
                eat(Tokens.ID.ordinal)
                eat(Tokens.SEMI.ordinal)
                varDeclaration()
            }
            Tokens.INT.ordinal->{
                eat(Tokens.INT.ordinal)
                eat(Tokens.ID.ordinal)
                eat(Tokens.SEMI.ordinal)
                varDeclaration()
            }
            Tokens.FLOAT.ordinal->{
                eat(Tokens.FLOAT.ordinal)
                eat(Tokens.ID.ordinal)
                eat(Tokens.SEMI.ordinal)
                varDeclaration()
            }
        }

    }


}